src/Oscoin/Crypto/Address.hs

```diff
@@ -0,0 +1,106 @@
+-- | An address format for the Oscoin blockchain loosely based on Parity's SS58
+-- and on BTC's Base58Check.
+--
+-- See @docs/address_format.md@ for an extensive documentation.
+--
+
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Oscoin.Crypto.Address
+    ( Address(addressPrefix, addressPayload)
+    , AddressPrefix
+    , AddressPayload
+
+    -- * Creating an Address
+    , fromPublicKey
+
+    -- * Parsing binary blobs back into addresses
+    , parseAddress
+
+    -- * Pretty-printing an Address
+    , renderAddress
+    , fmtAddress
+    ) where
+
+import           Oscoin.Prelude
+
+import qualified Oscoin.Configuration as Config
+import           Oscoin.Crypto.Address.Internal
+import           Oscoin.Crypto.Address.Serialisation as Serialisation
+import           Oscoin.Crypto.Hash
+import           Oscoin.Crypto.PubKey
+
+import qualified Codec.CBOR.ByteArray as CBOR
+import qualified Codec.Serialise as CBOR
+import           Data.ByteArray (ByteArrayAccess, convert)
+import           Data.ByteString.BaseN
+                 (decodeBase32z, encodeBase32z, encodedText)
+import           Data.Tagged
+import qualified Formatting as F
+
+{------------------------------------------------------------------------------
+    Creating an Address
+------------------------------------------------------------------------------}
+
+data AddressCreationError =
+    UnsupportedConfiguration Config.Network
+    deriving Show
+
+-- | Creates an 'Address' given a 'Network' configuration and a 'PublicKey'.
+fromPublicKey
+    :: forall c.
+       ( ByteArrayAccess (ShortHash c)
+       , Hashable c (PublicKey c)
+       )
+    => Config.Network
+    -- ^ The 'Network' the node is running on.
+    -> PublicKey c
+    -- ^ A 'PublicKey'.
+    -> Either AddressCreationError (Address c)
+fromPublicKey network pk = runExcept $
+    case configToAddressTypeTag network of
+      Nothing -> throwError $ UnsupportedConfiguration network
+      Just at -> do
+          let prefix = AddressPrefix
+                { addressFormat   = AddressFormat AddressFormatTagCBOR
+                , addressType     = AddressType at
+                , protocolVersion = ProtocolVersion ProtocolVersion_V1
+                }
+              payload = Tagged . CBOR.fromByteString . convert . shortHash @c $ pk
+          pure $ Address prefix (AddressPayload payload)
+
+-- | Parses a base32z-encoded binary block back into an 'Address'.
+parseAddress
```

Comment:
Should be `decodeAddress` if it's operating on binary data

Reply:
Probably more effective, yup.

Reply:
35cf933